base/dict.jl

```diff
@@ -351,8 +351,8 @@ ht_keyindex2!(h::Dict, key) = ht_keyindex2_shorthash!(h, key)[1]
 
     sz = length(h.keys)
     # Rehash now if necessary
-    if h.ndel >= ((3*sz)>>2) || h.count*3 > sz*2
-        # > 3/4 deleted or > 2/3 full
+    if (h.count + h.ndel)*3 > sz*2
```

Comment:
This is great!

If we did want to preserve the 3/4 (9/12th) vs. 2/3 (8/12th) difference, we could scale them separately:
```
(h.count*3/2 + h.ndel*4/3) > sz
```
==>
```
(h.count*9 + h.ndel*8) > sz*12
```
but I assume that is not worthwhile?

If useful, we could also trade space for more performance here too, by moving this computation into `rehash!`, and having an extra field that counts down to 0 and then triggers a rehash:
```
rehash_target::Int = length(h.slots) - 8 * h.count
# in delete: rehash_target::Int -= 8
# in setindex: rehash_target::Int -= 9 # - 8 (if it was a tombstone)
rehash_target < 0 && rehash!(h)
```

Reply:
Thanks. 

The difference between `h.count*9` and `h.ndel*8` seems to be negligible, thus, I don't see it worthwhile.

Using a counter is and interesting idea, but I don't believe in the performance improvement as the multiplication is probably not the bottleneck here. Moreover, sometimes the re-hashing may not be necessary if the size is not changing a lot.